### ISSUE TITLE
Further training of TTS models

### ISSUE THREAD

**Issue by Arafat4341** [espnet/espnet#1896]
Can we train further the models for TTS? Specially Tacotron-2 and parallel_wavegan

**Comment by Fhrozen** [comment]
Can you define `further`?
If you mean by retrain, you can check the details at https://github.com/espnet/espnet/issues/1889
In case you mean longer train means: then, in the recipe you are using, set `epochs: n` for `n` be the number you want to train for. 

As example, modify in this line if you are using libritts: https://github.com/espnet/espnet/blob/2652d1a61edb288f05203bca71d92afde9e6f3e7/egs/libritts/tts1/conf/train_pytorch_transformer%2Bspkemb.yaml#L63

**Comment by marcosnetopires** [comment]
If you are talking about training a previously trained model, then the previous answer was right (with the reference to #1889 ), but you can't retrain models with Phoneme inputs, only with char inputs.

You cant't use phoneme pre-trained models because new data will most likely have new phonemes that the original dataset did not have, so there will be more input values (phonemes) than input nodes in the model. With chars there is probably no problem as all chars are already used in the original dataset.

**Comment by Arafat4341** [comment]
For Japanese, I am trying to enhance the quality. Can we make the pre-trained model perform better? Is there any way?

**Comment by Arafat4341** [comment]
@Fhrozen Isn't there any way to increase training steps for JSUT?

**Comment by Fhrozen** [comment]
You mean by increasing the epochs (number of training steps). 

https://github.com/espnet/espnet/blob/39437adc7b93f174036110400f7b907249c6f628/egs/jsut/tts1/conf/train_pytorch_transformer.yaml#L63

?

**Comment by Arafat4341** [comment]
@Fhrozen Yeah

**Comment by Fhrozen** [comment]
the script for training? You mean: https://github.com/espnet/espnet/blob/master/egs/jsut/tts1/run.sh

You can execute it with:
```bash
$ ./run.sh --stage 3 --ngpu <your_number_of_gpus> --train_config conf/train_pytorch_transformer.yaml
```
Execute it after modifying `conf/train_pytorch_transformer.yaml` with the number of epochs.
The quality depends on several parameters: Number of samples in the training set (with or w/o augmentation), network mode, batchsize, type of loss, number of epochs,  and so on.
The script is already optimized so you can start training with these parameters and try to see the effects on changing any.

**Comment by Arafat4341** [comment]
@Fhrozen Thanks a lot!

**Comment by Arafat4341** [comment]
@marcosnetopires according to 1889, I tried to add --pretrained-model option while running https://github.com/espnet/espnet/blob/master/egs/jsut/tts1/run.sh.

I get `./run.sh: invalid option --pretrained-model`

@Fhrozen  I want to train from previously saved checkpoint from pre-trained model with extended training steps. I added option `--pretrained-model <location/of/pre-trained/model>` while running run.sh.

Getting the above mentioned error.

**Comment by Fhrozen** [comment]
Ok, there are few things you need to be careful about. Theres is no `--pretrained-model` in any script.
If you want to continue a training from a previous state, you can use `--resume` to do so in `run.sh`.
If you want to use as initial model (pretraining), then you need to add in the train.yaml file (espnet/egs/jsut/tts1/conf/train_pytorch_transformer.yaml) the flags 

```
enc-init: <your model> 
dec-init: <your model>
```
https://github.com/espnet/espnet/blob/39437adc7b93f174036110400f7b907249c6f628/espnet/bin/tts_train.py#L259-L276

You can check all the flags available in the previous file. A description for each one is also included.

**Comment by Arafat4341** [comment]
@Fhrozen Okay. Thanks a lot! Would you tell me how can I put the data on place. The `data.json` file. What is the script for that?!

**Comment by Fhrozen** [comment]
It should be prepared in stage 2 https://github.com/espnet/espnet/blob/master/egs/jsut/tts1/run.sh
`./run.sh --stage 2`.  in the folder: `dump/phn_train_no_dev`

**Comment by Arafat4341** [comment]
@Fhrozen Thanks a lot. Actually I am getting error in the `Stage: 2`

`/content/espnet/egs/jsut/tts1/../../../utils/data2json.sh: line 107: bc: command not found`

I don't understand why I am getting this.

**Comment by Fhrozen** [comment]
You need to install some programs using `sudo apt install <programs>`

You can take as reference the docker file to see which programs are required:
https://github.com/espnet/espnet/blob/8c33d3774ecbdef47485dadacef5edc032ef5f50/docker/prebuilt/runtime/Dockerfile#L11-L31

**Comment by Arafat4341** [comment]
@Fhrozen Thanks a lot! I had 2 programs missing.

Actually I want to train tacotron2 model.
SO I am using this command:
`./run.sh --stage -1 --ngpu 1 --train_config conf/train_pytorch_tacotron2.yaml --enc-init <tacotron2/model/path>/model.last1.avg.best`

But I am getting: `./run.sh: invalid option --enc-init`

I am wondering why is `--enc-init` invalid option!

**Comment by Fhrozen** [comment]
Is because this flag is not implemented in `run.sh`.  The only available flags for `run.sh` is https://github.com/espnet/espnet/blob/8c33d3774ecbdef47485dadacef5edc032ef5f50/egs/jsut/tts1/run.sh#L10-L51

To use pretraining, add `enc-init: <your-model>` inside the training file: `conf/train_pytorch_tacotron2.yaml`

**Comment by Arafat4341** [comment]
@Fhrozen , How to make it work?!
I added this line: `enc-init: "path/to/model.last1.avg.best" to train_pytorch_tacotron2.yaml.
And also added enc-init="" to `run.sh`.
Getting:
```
./run.sh: line 20: enc-init=: command not found
./run.sh: invalid option --enc-init
```

As far as I understand the `train_pytorch_tacotron2.yaml` file contains the hyper-parameters for model right? I was just wondering why we are adding `enc-init:` here!

**Comment by b-flo** [comment]
Hi,

You should pay more attention to what @Fhrozen said : 

> Is because this flag is not implemented in run.sh
>
> To use pretraining, add enc-init: <your-model> inside the training file: conf/train_pytorch_tacotron2.yaml

`enc-init` and `dec-init` are options for `tts_train.py` : https://github.com/espnet/espnet/blob/8c33d3774ecbdef47485dadacef5edc032ef5f50/egs/jsut/tts1/run.sh#L155-L166

The options are passed to it through the config file (L.166). You don't define them in `run.sh`.
